module/zfs/zvol.c

```diff
@@ -847,8 +850,10 @@ zvol_discard(void *arg)
 		size = end - start;
 	}
 
-	if (start >= end)
-		goto out;
+	if (start >= end) {
+		error = SET_ERROR(EIO);
```

Comment:
IIRC correctly there are cases when the kernel can submit zero length discards and set FUA.  In which case the function should only call `zil_commit()` (which was obviously missing).  So this should be updated to leave `error = 0`.  @tuxoko may better remember the exact details here.